strawberryfields/apps/graph/utils.py

```diff
@@ -0,0 +1,143 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+r"""
+Graph functions
+===============
+
+**Module name:** :mod:`strawberryfields.future.apps.graph.utils`
+
+.. currentmodule:: strawberryfields.future.apps.graph.utils
+
+This module provides some ancillary functions for dealing with graphs. This includes
+:func:`is_undirected` to check if an input matrix corresponds to an undirected graph (i.e.,
+is symmetric) and :func:`subgraph_adjacency` to return the adjacency matrix of a subgraph when an
+input graph and subset of nodes is specified.
+
+Furthermore, the frontend :func:`~strawberryfields.future.apps.graph.dense_subgraph.find_dense`
+function allows users to input graphs both as a `NumPy <https://www.numpy.org/>`__ array
+containing the adjacency matrix and as a `NetworkX <https://networkx.github.io/>`__ ``Graph``
+object. The :func:`to_networkx_graph` function allows both inputs to be processed into a NetworkX
+Graph for ease of processing.
+
+Summary
+-------
+
+.. autosummary::
+    is_undirected
+    subgraph_adjacency
+    to_networkx_graph
+    is_subgraph
+    update_options
+
+Code details
+------------
+"""
+from typing import Union, Iterable
+
+import networkx as nx
+import numpy as np
+
+graph_type = Union[nx.Graph, np.ndarray]
+
+
+def to_networkx_graph(graph: graph_type) -> nx.Graph:
+    """Converts input into a NetworkX ``Graph`` (i.e., an undirected graph).
+
+    Given an input undirected graph of type ``graph_type = Union[nx.Graph, np.ndarray]``,
```

Comment:
Minor comment, but within function docstrings, it might be more readable for less technical users to avoid the type-hint notation, and instead write something plain and simple like

```rest
    Converts an undirected NetworkX graph or real symmetric NumPy adjacency matrix
    into a NetworkX graph object.
```

The input and return types don't need to be explicitly mentioned here, as it is just duplicating the `Args:` and `Returns:` section.

Reply:
A follow on thought: it seems a bit weird to have this function also act on `nx.Graph` objects, but just perform the identity. I remember that this function was originally a convenience function to avoid code duplication, is it still needed with the current state of the codebase?

Might make more sense to position this function more for input validation of graphs (and conveniently also returning a NX graph) as opposed to a casting function.

Reply:
Nice, have updated now to `validate_graph` and updated docstring. We still do use this function as in `find_dense` we allow the user to put in either type of graph. If we remove the `find_dense` wrapper, we can then allow this dual input functionality in the wrapped functions.